File: Assets/Code/Logic/Hero/Model/HeroData.cs
using UnityEngine;
using System.Collections.Generic;
using Common.Util;
using Common.ResMgr;
using Logic.Enums;

namespace Logic.Hero.Model
{
    public class HeroData
    {
        private static Dictionary<int, HeroData> _heroDataDictionary;

        public static Dictionary<int, HeroData> GetHeroDatas()
        {
            if (_heroDataDictionary == null)
            {
                _heroDataDictionary = CSVUtil.Parse<int, HeroData>("config/csv/hero", "id");
            }
            return _heroDataDictionary;
        }

        public static Dictionary<int, HeroData> HeroDataDictionary
        {
            get
            {
                if (_heroDataDictionary == null)
                {
                    GetHeroDatas();
                }
                return _heroDataDictionary;
            }
        }

        public static HeroData GetHeroDataByID(int heroID)
        {
            HeroData heroData = null;
            if (HeroDataDictionary.ContainsKey(heroID) && HeroDataDictionary[heroID] != null)
            {
                heroData = HeroDataDictionary[heroID];
            }
            return heroData;
        }
		public bool IsPlayer()
		{
			return hero_type == 2;
		}
		
        [CSVElement("id")]
        public int id;

        [CSVElement("hero_type")]
        public int hero_type;

        [CSVElement("hitId")]
        public uint hitId;

        [CSVElement("skillId1")]
        public uint skillId1;

        [CSVElement("skillId2")]
        public uint skillId2;

        [CSVElement("passiveId1")]
        public uint passiveId1;

        [CSVElement("passiveId2")]
        public uint passiveId2;

        [CSVElement("name")]
        public string name;

        [CSVElement("description")]
        public string description;

        public string[] modelNames;
        public string defaultModelName;
        [CSVElement("model")]
        public string model
        {
            set
            {
                modelNames = value.Split(CSVUtil.SYMBOL_SEMICOLON);
                defaultModelName = modelNames[0];
            }
        }

        [CSVElement("height")]
        public float height;

        public ShadowType shadowType;
        [CSVElement("shadows")]
        public int shadowsInt
        {
            set
            {
                shadowType = (ShadowType)value;
            }
        }

        public Vector3 homeRotation;
        [CSVElement("home_rotation")]
        public string homeRotationString
        {
            set
            {
                homeRotation = value.ToVector3(CSVUtil.SYMBOL_SEMICOLON);
            }
        }

        public Vector3 rotation;
        [CSVElement("rotation")]
        public string rotationString
        {
            set
            {
                rotation = value.ToVector3(CSVUtil.SYMBOL_SEMICOLON);
            }
        }

        public Vector3 scale;
        [CSVElement("scale")]
        public string scaleString
        {
            set
            {
                scale = value.ToVector3(CSVUtil.SYMBOL_SEMICOLON);
            }
        }

        public string[] headIcons;
        [CSVElement("headIcon")]
        public string headIcon
        {
            set
            {
                headIcons = value.Split(CSVUtil.SYMBOL_SEMICOLON);
            }
        }

        [CSVElement("expMax")]
        public uint expMax;

        [CSVElement("expMax50")]
        public uint expMax50;

        [CSVElement("hp")]
        public uint HP;

        [CSVElement("hp_add")]
        public uint hpAdd;

        [CSVElement("normal_atk")]
        public uint normalAtk;

        [CSVElement("normal_atk_add")]
        public uint normalAtkAdd;

        [CSVElement("normal_def")]
        public uint normalDef;

        [CSVElement("normal_def_add")]
        public uint normalDefAdd;

        [CSVElement("magic_atk")]
        public uint magicAtk;

        [CSVElement("magic_atk_add")]
        public uint magicAtkAdd;

        [CSVElement("magic_def")]
        public uint magicDef;

        [CSVElement("magic_def_add")]
        public uint magicDefAdd;

        [CSVElement("speed")]
        public uint speed;

        [CSVElement("hit")]
        public float hit;

        [CSVElement("dodge")]
        public float dodge;

        [CSVElement("crit")]
        public float crit;

        [CSVElement("anti_crit")]
        public float antiCrit;

        [CSVElement("block")]
        public float block;

        [CSVElement("anti_block")]
        public float antiBlock;

        [CSVElement("counter_atk")]
        public float counterAtk;

        [CSVElement("crit_hurt_add")]
        public float critHurtAdd;

        [CSVElement("crit_hurt_dec")]
        public float critHurtDec;

        [CSVElement("armor")]
        public float armor;

        [CSVElement("damage_dec")]
        public float damageDec;

        [CSVElement("damage_add")]
        public float damageAdd;

        [CSVElement("starMin")]
        public uint starMin;

        [CSVElement("starMax")]
        public uint starMax;

        [CSVElement("correction")]
        public int correction;//战力修正值

        public RoleType roleType;
        [CSVElement("type")]
        public int type
        {
            set
            {
                roleType = (RoleType)value;
            }
        }

        [CSVElement("weaponType")]
        public uint weaponType;

        [CSVElement("floatable")]
        public bool floatable;

        public List<uint[]> bornEffectIds;
        [CSVElement("born_effect")]
        public string bornEffectIdStr
        {
            set
            {
                bornEffectIds = new List<uint[]>();
                string[] strs = value.ToArray(CSVUtil.SYMBOL_SEMICOLON);
                for (int i = 0, length = strs.Length; i < length; i++)
                {
                    uint[] effects = strs[i].ToArray<uint>(CSVUtil.SYMBOL_PIPE);
                    bornEffectIds.Add(effects);
                }
            }
        }

        [CSVElement("audio_attack")]
        public int audioAttack;

        [CSVElement("audio_under_attack")]
        public int audioUnderAttack;

        [CSVElement("audio_die")]
        public int audioDie;

        [CSVElement("effect_die")]
        public int dieEffectId;

		public RoleQuality roleQuality = RoleQuality.White;
		[CSVElement("quality")]
		public int quality
		{
			set
			{
				roleQuality = (RoleQuality)value;
			}
		}
		[CSVElement("quality_attr")]
		public float quality_attr;
		public string NameWithQualityColor
		{
			get
			{
				return UI.UIUtil.FormatStringWithinQualityColor(this.roleQuality, Common.Localization.Localization.Get(this.name));
			}
		}
    }
}

File: Assets/ToLua/Lua/user/ui/hero/model/hero_data.lua
local t = {}
local item = {}
item.__index = item

function item.New(table)
	local o = {}
	setmetatable(o,item)
	o.id = tonumber(table.id)
	o.hero_type = tonumber(table.hero_type) -- 区别英雄、主角、boss
	o.hitId = tonumber(table.hitId)
	o.skillId1 = tonumber(table.skillId1)
	o.skillId2 = tonumber(table.skillId2)
	o.passiveId1 = tonumber(table.passiveId1)
	o.passiveId2 = tonumber(table.passiveId2)
	o.name = table.name
	o.description = table.description
  o.modelNames = string.split(table.model,';')
  
	o.height = tonumber(table.height)
	o.shadows = tonumber(table.shadows)
  local rot = string.split2number(table.home_rotation,';')
	o.home_rotation = Vector3(rot[1],rot[2],rot[3])
	o.rotation = table.rotation
	o.scale = table.scale
	o.headIcons = string.split(table.headIcon,';')
	o.expMax = tonumber(table.expMax)
	o.expMax50 = tonumber(table.expMax50)
	o.hp = tonumber(table.hp)
	o.hp_add = tonumber(table.hp_add)
	o.normal_atk = tonumber(table.normal_atk)
	o.normal_atk_add = tonumber(table.normal_atk_add)
	o.magic_atk = tonumber(table.magic_atk)
	o.magic_atk_add = tonumber(table.magic_atk_add)
	o.normal_def = tonumber(table.normal_def)
	o.normal_def_add = tonumber(table.normal_def_add)
	o.magic_def = tonumber(table.magic_def)
	o.magic_def_add = tonumber(table.magic_def_add)
	o.speed = tonumber(table.speed)
	o.hit = tonumber(table.hit)
	o.dodge = tonumber(table.dodge)
	o.crit = tonumber(table.crit)
	o.anti_crit = tonumber(table.anti_crit)
	o.block = tonumber(table.block)
	o.anti_block = tonumber(table.anti_block)
	o.counter_atk = tonumber(table.counter_atk)
	o.crit_hurt_add = tonumber(table.crit_hurt_add)
	o.crit_hurt_dec = tonumber(table.crit_hurt_dec)
	o.armor = tonumber(table.armor)
	o.damage_dec = table.damage_dec
	o.damage_add = table.damage_add
	o.starMin = tonumber(table.starMin)
	o.starMax = tonumber(table.starMax)
	o.correction = tonumber(table.correction)
	o.roleType = tonumber(table.type)
	o.weaponType = tonumber(table.weaponType)
	o.floatable = tonumber(table.floatable)
  
  local heroBaseEquipIDStrs = string.split(table.hero_equment, ';')
  o.basicWeaponID = tonumber(heroBaseEquipIDStrs[1])
  o.basicArmorID = tonumber(heroBaseEquipIDStrs[2])
  o.basicAccessoryID = tonumber(heroBaseEquipIDStrs[3])
  
  o.quality = tonumber(table.quality)
  o.quality_attr = tonumber(table.quality_attr)
  o.quality_icon = table.quality_icon
	return o
end
function item:IsPlayer()
  return self.hero_type == 2
end

function item:GetNameWithQualityColor ()
  return ui_util.FormatStringWithinQualityColor(self.quality, LocalizationController.instance:Get(self.name))
end

function item:GetRankImagePath ()
  return 'sprite/main_ui/'..self.quality_icon
end

function item:GetRankRoleTypeSprite ()
  local ui_util = require('util/ui_util')
  return ui_util.GetRankRoleTypeSprite (self.roleType)
end

function t.GetDataById(id)
	return t[id]
end



local function Start()
	local origin = dofile('hero')
	origin.ForEach(function(id,table)
		local newItem = item.New(table)
		t[id] = newItem
	end)
end

Start()
return t